hail/python/hail/methods/impex.py

```diff
@@ -3206,8 +3218,48 @@ def add_reference_allele(mt):
 
     vd = vd.annotate_rows(filters=hl.coalesce(site[vd.locus].filters, hl.empty_set(hl.tstr)))
 
-    vd = vd.annotate_rows(as_vqsr = hl.dict(vqsr.index(vd.locus, all_matches=True).map(lambda record: (record.alt, record.drop('ref', 'alt')))))
-    vd = vd.annotate_globals(tranche_data=tranche.collect(_localize=False))
+    vd = vd.annotate_rows(as_vqsr = hl.dict(vqsr.index(vd.locus, all_matches=True)
+                                            .map(lambda record: (record.alt, record.drop('ref', 'alt')))))
+    vd = vd.annotate_globals(tranche_data=tranche.collect(_localize=False),
+                             truth_sensitivity_snp_threshold=truth_sensitivity_snp_threshold,
+                             truth_sensitivity_indel_threshold=truth_sensitivity_indel_threshold)
+
+    sorted_tranche_data = hl.sorted(vd.tranche_data, key=lambda x: x.truth_sensitivity)
+    vd = vd.annotate_globals(snp_vqslod_threshold=
+                             sorted_tranche_data.filter(lambda x: (x.model == 'SNP') & (
+                                         x.truth_sensitivity >= truth_sensitivity_snp_threshold))
+                             .head()
+                             ,
+                             indel_vqslod_threshold=sorted_tranche_data.filter(lambda x: (x.model == 'INDEL') & (
+                                         x.truth_sensitivity >= truth_sensitivity_indel_threshold))
+                             .head()
+                             )
+
+    vd = vd.annotate_rows(
+        allele_NO=vd.alleles[1:].map(
+            lambda allele: hl.coalesce(vd.as_vqsr.get(allele).yng_status == 'N', False)),
+        allele_YES=vd.alleles[1:].map(
+            lambda allele: hl.coalesce(vd.as_vqsr.get(allele).yng_status == 'Y', True)),
+        allele_OK=vd.alleles[1:].map(
+            lambda alt: hl.coalesce(vd.as_vqsr.get(alt).vqslod > hl.if_else(hl.is_snp(vd.alleles[0], alt),
+                                                                            truth_sensitivity_snp_threshold,
+                                                                            truth_sensitivity_indel_threshold),
+                                    True)),
+    )
+
+    ft_criteria = (hl.range(vd.LGT.ploidy)
+                   .map(lambda idx: vd.LA[vd.LGT[idx]])
+                   .filter(lambda x: x != 0)
+                   .fold(lambda tuple, called_idx: hl.struct(
+        any_no=tuple[0] | vd.allele_NO[called_idx-1],
+        any_yes=tuple[1] | vd.allele_YES[called_idx-1],
+        all_ok_vqslod=tuple[2] & vd.allele_OK[called_idx-1])
+    , hl.struct(any_no=False, any_yes=False, all_ok_vqslod=False)))
```

Comment:
Instead of True/False/NA, would it be possible to do FAIL/PASS/. ?

Reply:
Replacing an entry-level boolean (1 byte in Hail representations) with a string is going to be quite expensive. The gnomAD team has used filters this way (boolean fields) in the past and has had few problems.

Reply:
iiiinteresting. Okay, this is helpful to know! And maybe this would mean adding the "culprit" string would be very expensive as well

Reply:
Yes, though we could use an integer to indicate state (we've been meaning to add enumeration types but haven't had a chance, that would make it easier).